doc/devel/contributing.rst

```diff
@@ -0,0 +1,422 @@
+  code.  The interface code should be named `FOO_wrap.cpp` or
+  `FOO_wrapper.cpp`.
+
+* Header file documentation (aka docstrings) should be in Numpydoc
+  format.  We don't plan on using automated tools for these
+  docstrings, and the Numpydoc format is well understood in the
+  scientific Python community.
+
+.. _keyword-argument-processing:
+
+Keyword argument processing
+---------------------------
+
+Matplotlib makes extensive use of ``**kwargs`` for pass-through
+customizations from one function to another.  A typical example is in
+:func:`matplotlib.pylab.text`.  The definition of the pylab text
+function is a simple pass-through to
+:meth:`matplotlib.axes.Axes.text`::
+
+  # in pylab.py
+  def text(*args, **kwargs):
+      ret =  gca().text(*args, **kwargs)
+      draw_if_interactive()
+      return ret
+
+:meth:`~matplotlib.axes.Axes.text` in simplified form looks like this,
+i.e., it just passes all ``args`` and ``kwargs`` on to
+:meth:`matplotlib.text.Text.__init__`::
+
+  # in axes.py
+  def text(self, x, y, s, fontdict=None, withdash=False, **kwargs):
+      t = Text(x=x, y=y, text=s, **kwargs)
+
+and :meth:`~matplotlib.text.Text.__init__` (again with liberties for
+illustration) just passes them on to the
+:meth:`matplotlib.artist.Artist.update` method::
+
+  # in text.py
+  def __init__(self, x=0, y=0, text='', **kwargs):
+      Artist.__init__(self)
+      self.update(kwargs)
+
+``update`` does the work looking for methods named like
+``set_property`` if ``property`` is a keyword argument.  i.e., no one
+looks at the keywords, they just get passed through the API to the
+artist constructor which looks for suitably named methods and calls
+them with the value.
+
+As a general rule, the use of ``**kwargs`` should be reserved for
+pass-through keyword arguments, as in the example above.  If all the
+keyword args are to be used in the function, and not passed
+on, use the key/value keyword args in the function definition rather
+than the ``**kwargs`` idiom.
+
+In some cases, you may want to consume some keys in the local
+function, and let others pass through.  You can ``pop`` the ones to be
+used locally and pass on the rest.  For example, in
+:meth:`~matplotlib.axes.Axes.plot`, ``scalex`` and ``scaley`` are
+local arguments and the rest are passed on as
+:meth:`~matplotlib.lines.Line2D` keyword arguments::
+
+  # in axes.py
+  def plot(self, *args, **kwargs):
+      scalex = kwargs.pop('scalex', True)
+      scaley = kwargs.pop('scaley', True)
+      if not self._hold: self.cla()
+      lines = []
+      for line in self._get_lines(*args, **kwargs):
+          self.add_line(line)
+          lines.append(line)
+
+Note: there is a use case when ``kwargs`` are meant to be used locally
+in the function (not passed on), but you still need the ``**kwargs``
+idiom.  That is when you want to use ``*args`` to allow variable
+numbers of non-keyword args.  In this case, python will not allow you
```

Comment:
And we mirrored that API very closely at the beginning.  There is probably a few more cases (but I do not know any off the top of my head).

This should have a comment added that this is only a python2 restriction, as soon as we drop 2.7 as supported we can use keyword-only arguments :smile: (that is the python3 feature I like the most).